src/server.py

```diff
@@ -0,0 +1,188 @@
+            server_unit_databag: The relation databag content from agent relation.
+            unit_name: The agent unit name.
+
+        Returns:
+            Metadata if complete values(url, secret) are set. None otherwise.
+        """
+        address = server_unit_databag.get("url")
+        secret = server_unit_databag.get(f"{unit_name}_secret")
+        if not address or not secret:
+            return None
+        return Credentials(address=address, secret=secret)
+
+
+def get_credentials(
+    relation_name: str, unit_name: str, databag: typing.Optional[ops.RelationDataContent]
+) -> typing.Optional["Credentials"]:
+    """Get credentials from databag.
+
+    Args:
+        relation_name: The relation name of the databag. Either "slave" or "agent".
+        unit_name: The agent unit name.
+        databag: The relation databag for given relation.
+
+    Returns:
+        Credentials if databag contains valid metadata. None if partial or no metadata found.
+    """
+    if databag is None:
+        return None
+    if relation_name == state.SLAVE_RELATION:
+        return Credentials.from_jenkins_slave_interface_dict(databag)
+    return Credentials.from_jenkins_agent_v0_interface_dict(databag, unit_name=unit_name)
+
+
+class ServerBaseError(Exception):
+    """Represents errors with interacting with Jenkins server."""
+
+
+class AgentJarDownloadError(ServerBaseError):
+    """Represents an error downloading agent JAR executable."""
+
+
+def download_jenkins_agent(server_url: str, connectable_container: ops.Container) -> None:
+    """Download Jenkins agent JAR executable from server.
+
+    Args:
+        server_url: The Jenkins server URL address.
+        connectable_container: The connectable agent container.
+
+    Raises:
+        AgentJarDownloadError: If an error occurred downloading the JAR executable.
+    """
+    try:
+        res = requests.get(f"{server_url}/jnlpJars/agent.jar", timeout=300)
+        res.raise_for_status()
+    except (requests.HTTPError, requests.Timeout, requests.ConnectionError) as exc:
+        logger.error("Failed to download agent JAR executable from server, %s", exc)
+        raise AgentJarDownloadError(
+            "Failed to download agent JAR executable from server."
+        ) from exc
+
+    connectable_container.push(
+        path=AGENT_JAR_PATH, make_dirs=True, source=res.content, user=USER, group=GROUP
+    )
+
+
+def validate_credentials(
+    agent_name: str, credentials: Credentials, connectable_container: ops.Container
+) -> bool:
+    """Check if the credentials can be used to register to the server.
+
+    Args:
+        agent_name: The Jenkins agent name.
+        credentials: Server credentials required to register to Jenkins server.
+        connectable_container: The Jenkins agent workload container.
+
+    Returns:
+        True if credentials and agent_name pairs are valid, False otherwise.
+    """
+    # IMPORTANT: add random jitter to prevent parallel execution.
```

Comment:
Perhaps `random delay`? jitter is a networking concept I'm not sure is applicable?

Reply:
Renamed, thank you!